HAL/keymaster/4.1/JavacardKeymaster4Device.cpp

```diff
@@ -1071,78 +1025,160 @@ Return<void> JavacardKeymaster4Device::begin(KeyPurpose purpose, const hidl_vec<
+    hidl_vec<KeyParameter>& outParams, uint64_t& operationHandle,
+    OperationType& operType) {
+    ErrorCode errorCode = ErrorCode::UNKNOWN_ERROR;
+    if (operType == OperationType::PUBLIC_OPERATION) {
+        errorCode = handleBeginPublicKeyOperation(purpose, keyBlob, inParams,
+                                                  outParams, operationHandle);
+
+        // For Symmetric operations handleBeginPublicKeyOperation function
+        // returns INCOMPATIBLE_ALGORITHM error. Based on this error
+        // condition it fallbacks to private key operation.
+        if (errorCode == ErrorCode::INCOMPATIBLE_ALGORITHM) {
+            operType = OperationType::PRIVATE_OPERATION;
+        }
     }
-    /* Create a new operation handle and add a entry inside the operation table map with
-     * key - new operation handle
-     * value - hal generated operation handle.
+
+    if (operType == OperationType::PRIVATE_OPERATION) {
+        errorCode = handleBeginPrivateKeyOperation(
+            purpose, keyBlob, inParams, authToken, outParams, operationHandle);
+    }
+
+    // check if operation handle exists.
+    if (ErrorCode::OK == errorCode && isOperationHandleExists(operationHandle)) {
+        LOG(DEBUG) << "Operation handle " << operationHandle << "already exists"
+            "in the opertion table. so aborting this opertaion.";
+        // abort the operation.
+        errorCode = abortOperation(
+            operationHandle, (operType == OperationType::PRIVATE_OPERATION));
+    }
+    return errorCode;
+}
+
+Return<void> JavacardKeymaster4Device::begin(KeyPurpose purpose,
+                                             const hidl_vec<uint8_t>& keyBlob,
+                                             const hidl_vec<KeyParameter>& inParams,
+                                             const HardwareAuthToken& authToken,
+                                             begin_cb _hidl_cb) {
+    ErrorCode errorCode = ErrorCode::UNKNOWN_ERROR;
+    uint64_t operationHandle = 0;
+    OperationType operType = OperationType::PRIVATE_OPERATION;
+    hidl_vec<KeyParameter> outParams;
+    LOG(DEBUG) << "INS_BEGIN_OPERATION_CMD purpose: " << (int32_t)purpose;
+    /*
+     * Asymmetric public key operations are processed inside softkeymaster and private
+     * key operations are processed inside strongbox keymaster.
+     * All symmetric key operations are processed inside strongbox keymaster.
+     * If the purpose is either ENCRYPT / VERIFY then the operation type is set
+     * to public operation and in case if the key turned out to be a symmetric key then
+     * handleBeginOperation() function fallbacks to private key operation.
      */
+    if (KeyPurpose::ENCRYPT == purpose || KeyPurpose::VERIFY == purpose) {
+        operType = OperationType::PUBLIC_OPERATION;
+    }
+    int retryCount = 0;
+    int maxRetries = 2;
+    for(; retryCount < maxRetries; retryCount++) {
```

Comment:
@BKSSMVenkateswarlu  I am not finding this loop logic interesting with break. Can do the following as we are only going to retry once.  

handleBeginOperation();
if(errorCode!=OK) 
      return with erorr.
else if (isOperationHandleExists()
      handleBeginOperation
     if(errorCode!=OK) return with error
    else if (operationHandleExists())
    return with unknown error.

Baiscally, I want to make it clear that retry is only going to happen in the situation of operationHandleConflict and nothing else..

Reply:
Updated the code as per your comments.